rasa/core/processor.py

```diff
@@ -742,6 +745,9 @@ async def _run_prediction_loop(
                     self.on_circuit_break(tracker, output_channel, self.nlg)
                 break
 
+            if prediction.is_end_to_end_prediction:
```

Comment:
do we have a test which makes sure that we run this / to avoid regression?

Reply:
You're right, I need to add a test for this part: I'm amending the `e2e_bot` domain file to contain slots, hope that's fine 🤞🏼 The test I've added passes, we'll see what the CI says 😅

Reply:
@wochinge Adding slots to the `e2e_bot` domain causes this test `test_e2e_with_entity_evaluation` to [fail](https://github.com/RasaHQ/rasa/runs/4167179517?check_suite_focus=true), because `report` variable is empty. Do you think it's better for me to create a different e2e fixture?

Reply:
I'd rather investigate why the report is empty to be honest.

Reply:
Ok, I deep dived in the area related to testing stories and this is what i found:

1. the result of [`evaluate_entities`](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/core/test.py#L1119) is an empty dict `{ }` for `report` [here](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/nlu/test.py#L882) because this [if statement](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/model_testing.py#L359) in `model_testing.py` gets executed.

2. Looking more into this: `entity_results` output [here](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/core/test.py#L1079) is an empty list -> because [`tracker_entity_results`](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/core/test.py#L955) is empty list -> because every `entity_result` [here](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/core/test.py#L838) is predicted as `None`.  

And this is because [this block](https://github.com/RasaHQ/rasa/blob/4620db7607d61baa050d9779b1c1264568ccef61/rasa/core/test.py#L604) never gets executed and I'm not sure why?

Reply:
I've looked at what's in `tracker.events` and based on the fact that a `SlotSet` event gets added now to the tracker after every `UserUttered`, that block in `_get_e2e_entity_evaluation_result` never gets executed because every last event will be a `SlotSet`, should I modify this method to actually get the penultimate event if last event is a `SlotSet`, test passes when I check that 🤔 :

```
['action_listen', 'UserUttered(text: my name is joe, intent: None, entities: joe (Type: name, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: name, value: joe)']
['action_listen', 'UserUttered(text: my name is joe, intent: None, entities: joe (Type: name, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: name, value: joe)', 'hi joe!']
['action_listen', 'UserUttered(text: my name is joe, intent: None, entities: joe (Type: name, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: name, value: joe)', 'hi joe!', 'action_listen', 'UserUttered(text: I am feeling happy, intent: None, entities: happy (Type: mood, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: mood, value: happy)']
['action_listen', 'UserUttered(text: my name is joe, intent: None, entities: joe (Type: name, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: name, value: joe)', 'hi joe!', 'action_listen', 'UserUttered(text: I am feeling happy, intent: None, entities: happy (Type: mood, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: mood, value: happy)', "I'm glad"]
['action_listen', 'UserUttered(text: today I was very cranky, intent: None, entities: cranky (Type: mood, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: mood, value: cranky)']
['action_listen', 'UserUttered(text: today I was very cranky, intent: None, entities: cranky (Type: mood, Role: None, Group: None), use_text_for_featurization: True)', 'SlotSet(key: mood, value: cranky)', "I'm sorry!"]
```

Reply:
@wochinge that seemed to have fixed the test 😁 Let me know what you think, ready for a new round of review.